src/docfx/build/document/DocumentProvider.cs

```diff
@@ -264,12 +264,53 @@ private static string PathToRelativeUrl(string path, ContentType contentType, Ur
     }
 
     /// <summary>
-    /// In docs, canonical URL is later overwritten by template JINT code.
-    /// TODO: need to handle the logic difference when template code is removed.
+    /// The logic is copied from template JINT code
     /// </summary>
-    private string GetCanonicalUrl(string siteUrl)
+    private string GetCanonicalUrl(string? sitePath)
     {
-        return $"https://{_config.HostName}/{_buildOptions.Locale}{siteUrl}";
+        if (sitePath == null)
+        {
+            return "";
+        }
+
+        var canonicalUrlPrefix = UrlUtility.Combine($"https://{_config.HostName}", _buildOptions.Locale, _config.BasePath);
+
+        var canonicalUrl = canonicalUrlPrefix + "/" + RemoveExtension(EncodePath(sitePath));
+        canonicalUrl = canonicalUrl.ToLowerInvariant();
+
+        if (canonicalUrl.EndsWith("/index"))
```

Comment:
It seems this part of logic is already handled by https://github.com/dotnet/docfx/blob/3f0d11c66cc11ce28597f668040e792fd7fcec7c/src/docfx/build/document/DocumentProvider.cs#L241-L241. Any chance to reuse original signature of `siteUrl`?

Reply:
If the only concern is encoding, please consider including the encoding logic in the generator of `siteUrl`.

Reply:
code updated, please review again